src/libstd/collections/hash/set.rs

```diff
@@ -677,8 +675,14 @@ impl<T, S> Extend<T> for HashSet<T, S>
           S: BuildHasher
 {
     fn extend<I: IntoIterator<Item = T>>(&mut self, iter: I) {
-        for k in iter {
-            self.insert(k);
+        // Keys may be already present or show multiple times in the iterator!
+        // Reserve only half the hint (round up), this guarantees the table
+        // will only resize twice in the worst case.
+        let iter = iter.into_iter();
+        let hint = iter.size_hint().0;
+        self.reserve((hint / 2) + (hint & 1));
+        for (k, v) in iter {
```

Comment:
This will not compile :P 

Might be simpler to just call down to `self.map.extend(iter.into_iter().map(|k| (k, ()))`.